interop_task/interop_task.md

```diff
@@ -0,0 +1,124 @@
+
+## Motivation
+
+SYCL does not allow a user to access cl_mem object out of an cl::sycl::accessor, it difficult/impossible to integrate OpenCL library to use them as is inside the data-flow execution model of SYCL, as the only current way to do this is to create all OpenCL buffers up-front, which is not always possible.
+
+This proposal introduces a way for a user to retrieve the OpenCL buffer associate with a SYCL buffer and enqueue a host task that can execute an arbitrary portion of host code within the SYCL runtime, therefore taking advantage of SYCL dependency analysis and scheduling.
+
+## Accessing low-level API functionality on SYCL queues
+
+We introduce a new type of handler, the **codeplay::handler**, which includes a new
+**interop\_task** method that enables submission of low-level API code from the host.
+By submitting this command group to the SYCL device queue, we guarantee it is
+executed in-order w.r.t the other command groups on the same queue.
+Simultaneously, we guarantee that this operation is performed
+asynchronously w.r.t to the user-thread (therefore, enabling the user
+thread to continue submitting command groups).
+Other command groups enqueued in the same or different queues
+can be executed following the sequential consistency by guaranteeing the
+satisfaction of the requisites of this command group.
+It is the user's responsability to ensure the lambda submitted via interop_task does not create race conditions with other command groups or with the host.
+
+The possibility of enqueuing host tasks on SYCL queues also enables the
+runtime to perform further optimizations when available.
+For example, a SYCL runtime may decide to map / unmap instead of copy operations,
+or performing asynchronous transfers while data is being computed.
+
+### cl::sycl::codeplay::handler
+
+```cpp
+namespace cl {
+namespace sycl {
+namespace codeplay {
+
+class handler : public cl::sycl::handler {
+ private:
+  // implementation defined constructor
+  handler(__unspecified__);
+
+ public:
+  /* Submit a task with interoperability statements. */
+  template <typename FunctorT>
+  void interop_task(FunctorT hostFunction);
+};
+}  // namespace codeplay
+}  // namespace sycl
+}  // namespace cl
+```
+
+### codeplay::handler::interop_task
+
+The `interop_task` allows users to submit tasks containing C++ statements with low-level API call (e.g. OpenCL Host API entries).
+The command group that encapsulates the task will execute following the usual SYCL dataflow execution rules.
+The functor passed to the `interop_task` takes as input a const reference to a `cl::sycl::codeplay::interop_handle`. The handle can be used to retrieve underlying OpenCL objects relative to the execution of the task.
+
+It is not allowed to allocate new SYCL object inside an `interop_task`.
+It is the user responsibilities to ensure all operations peroformed inside the `interop_task` finished before returning from it.
+
```

Comment:
Please add:
```
Although the statements inside the lambda submitted o the interop_task are executed on the host, the requirements and actions for the command group are satisied for the device.
This is the opposite of the host_handler vendor extension, where requisites are satisfied for the host since the statements on the lambda submited to the single task are meant to have side effects on the host only.
The interop task lambda can have side effects on the host, but it is the programmer responsability to ensure requirements dont need to be satisfied for the host.
```